docs/core/exceptions.rst

```diff
@@ -97,6 +96,34 @@ Error Classes
    :members:
    :show-inheritance:
 
+.. _error_subdocuments:
+
+ErrorSubdocuments
+-----------------
+
+Errors returned from APIs may define a series of subdocuments, each containing
+an error object. This is used if there were multiple errors encountered and the
+API wants to send them back all at once.
+
+All instances of ``GlobusAPIError`` define an attribute, ``errors``, which is
+an array of ``ErrorSubdocument``\s. It may be empty if there were no subdocuments.
```

Comment:
It is still possible in a small selection of cases.

1. `errors` is initialized to `[]`. If the body isn't JSON or is the wrong top-level shape (e.g., top-level array), `errors` will stay `[]`
2. A Type Zero or undefined format can produce an errors array of `[]` if it explicitly sets`"errors": []`.

Regarding (2), we talked about this and I was waffling about it. But in another conversation, we looked again at the draft error spec and the current wording there is as follows:

```
Clients MAY ignore any `"errors"` key in Type Zero Globus Errors and always set `errors` to `[]`.

If the error payload has a top-level key `"errors"` containing an array, and every element of that array is an object, then errors on the parsed response MAY be populated with the `"errors"` in the response.

Absent an `"errors"` key, clients MAY populate `errors` with the root response object.

Clients MUST NOT populate the `errors` of a Parsed Globus Error from a Type Zero Globus Error with data which is not here specified as a valid parse.
```

Now, obviously we're talking about a draft spec that we're working on contemporaneously with this first implementation. So we could change the wording.

But as specified here, the spec was written to say that `errors=[<root_doc>]` is valid only if `"errors"` was not present as a key. If it is present, our choices are:
- respect any well-formed `errors` data (to the above definition of well-formed-ness)
- set `errors=[]`

I'm not sure it's convincing and perfect but it makes some sense in that an API which says `errors=[]` has explicitly told us that there are no sub-errors.

Reply:
I'm just trying to make sure our documentation reflects reality of what is in this change.

The line "It may be empty if there were no subdocuments" makes me expect that

```
{
  "code": "BadRequest",
  "message": "it was a bad request",
  "request_id": "b1ce01db-0b2b-486b-9602-5ddaf0351ad9"
}
```

Would be deserialized into an object with an empty "errors" (there were no subdocuments)

---

As I understand it, this is not the case though.

Am I misunderstanding the implementation? Reading that line unfairly?